x/erc20/keeper/genesis.go

```diff
@@ -0,0 +1,152 @@
+// Copyright Tharsis Labs Ltd.(Evmos)
+// SPDX-License-Identifier:ENCL-1.0(https://github.com/evmos/evmos/blob/main/LICENSE)
+package keeper
+
+import (
+	"fmt"
+	"math/big"
+	"strings"
+
+	errorsmod "cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	"github.com/ethereum/go-ethereum/common"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/evmos/evmos/v16/ibc"
+	"github.com/evmos/evmos/v16/x/erc20/types"
+	"github.com/evmos/evmos/v16/x/evm/statedb"
+)
+
+// SetGenesisTokenPairs stores in state the provided token pairs in genesis
+// A genesis token pair is a token pair that will be registered from genesis.
+// A current use case of this is e2e tests,
+// where we may need to register a token pair to make it work.
+// Currently, there's no other way to do this because the RegisterCoinProposal was deprecated.
+func (k Keeper) SetGenesisTokenPairs(ctx sdk.Context, pairs []types.TokenPair) error {
+	if len(pairs) == 0 {
+		return nil
+	}
+	// We need to store the ERC20 contracts in the provided addresses at genesis.
+	// To do so, we'll need to seed the stateDB with those contracts.
+	// We cannot use EVM keeper functions for this because that will create a contract address
+	// dependent on the sender nonce.
+	stateDB := statedb.New(ctx, k.evmKeeper, statedb.TxConfig{})
+	for _, pair := range pairs {
+		contractAddr := pair.GetERC20Contract()
+		coinMeta, err := k.getTokenPairMeta(ctx, pair)
+		if err != nil {
+			return fmt.Errorf("error while generating metadata for genesis pair denom %s. %w", pair.Denom, err)
+		}
+		code, err := k.generateContractCode(ctx, stateDB, coinMeta, contractAddr)
+		if err != nil {
+			return fmt.Errorf("error while getting contract code for genesis pair denom %s. %w", pair.Denom, err)
+		}
+
+		// Store the ERC20 contract code in the address provided in genesis
+		stateDB.SetCode(contractAddr, code)
+
+		id := pair.GetID()
+		k.SetTokenPair(ctx, pair)
+		k.SetDenomMap(ctx, pair.Denom, id)
+		k.SetERC20Map(ctx, contractAddr, id)
+	}
+
+	// Commit the changes to effectively seed the state DB with
+	// the genesis token pairs ERC20 contracts
+	return stateDB.Commit()
+}
+
+// getTokenPairMeta is a helper function to generate token pair metadata for the genesis token pairs
+func (k Keeper) getTokenPairMeta(ctx sdk.Context, pair types.TokenPair) (banktypes.Metadata, error) {
+	// The corresponding IBC denom trace should be included in genesis
+	denomTrace, err := ibc.GetDenomTrace(*k.transferKeeper, ctx, pair.Denom)
+	if err != nil {
+		return banktypes.Metadata{}, err
+	}
+	// validate base denom length
+	if len(denomTrace.BaseDenom) < 2 {
+		return banktypes.Metadata{}, fmt.Errorf("denom trace base denom is too short. Should be at least 2 characters long, got '%s'", denomTrace.BaseDenom)
+	}
```

Comment:
shouldn't it be at least 5 chars for ibc coins? 🤔 aren't they all like `ibc/...`?

Reply:
the base denom is the native name of the coin, e.g. 'uosmo', 'aevmos', 'ustake', etc